tapcli: expose universe sync config settings

jharveyb opened the pull request:
Fixes https://github.com/lightninglabs/taproot-assets/issues/566.

In draft as still needs some manual testing.

Would also like to add a universe federation config info command before merge.

Intended supported usage would be to set the insert or export bools independently (one or both at the same time), and only modify the config for a single scope at a time (global config or config for a specific asset).

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -8,7 +8,6 @@ import (
 	tap "github.com/lightninglabs/taproot-assets"
 	"github.com/lightninglabs/taproot-assets/proof"
 	"github.com/lightninglabs/taproot-assets/taprpc"
-	"github.com/lightninglabs/taproot-assets/taprpc/universerpc"
```
Weird that linter didn't complain with both instances...

jharveyb replied:
Yeah updated it to follow the abbrevation we use elsewhere in `cmd`.

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -74,7 +73,11 @@ var universeRootsCommand = cli.Command{
 
-func parseUniverseID(ctx *cli.Context, mustParse bool) (*universerpc.ID, error) {
+func parseProofType(ctx *cli.Context) (*unirpc.ProofType, error) {
+	if !ctx.IsSet(proofTypeName) {
```
IIUC we set a default for issuance above, it always _should_ be set?

jharveyb replied:
Ah that came in after a rebase, will remove.

jharveyb replied:
Moving to no default for these commands to match the other flags.

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -698,6 +715,218 @@ func universeFederationDel(ctx *cli.Context) error {
+		err       error
+	)
+
+	// The cli.BoolFlag works for variables that are false by default,
```
I think we can use `BoolTFlag` here? If we want it to be true by default.

jharveyb replied:
I think we don't want it to have a default at all and have it be fully explicit.

Something like `tapcli universe federation config global --proof_type=transfer --insert true --export false`

Where `insert` and `export` without a bool fail.

Roasbeef replied:
Ah sure I think that makes sense.

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -698,6 +715,218 @@ func universeFederationDel(ctx *cli.Context) error {
+			Usage: "if true, remote Universes can push proofs to" +
+				"the local Universe",
+		},
+		cli.StringFlag{
```
Why can't this be a `bool`? Or is that we want to actually allow it to not bet set (so you don't have to always set both)?

jharveyb replied:
Also that, I think it would make the most sense to be able to set either field or both with one command, vs one command per field.

jharveyb replied:
Also for `ctx.BoolFlag`, actually trying to resolve the flag returns `false` if parsing fails, instead of a separate error.

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -698,6 +715,218 @@ func universeFederationDel(ctx *cli.Context) error {
+}
+
+var (
+	proofInsertName        = "insert"
```
`allow_insert` and `allow_export`?

Roasbeef commented on cmd/tapcli/universe.go:
```diff
@@ -698,6 +715,218 @@ func universeFederationDel(ctx *cli.Context) error {
+	}
+
+	if ctx.IsSet(proofExportName) {
+		*exportOpt, err = isValidBool(ctx.String(proofExportName))
```
Unclear why we need special proof handling here.

jharveyb replied:
Not sure what you meant?

Roasbeef replied:
Nvm, makes sense re decision to force a user to specify everything.

Roasbeef reviewed (commented):
Looks pretty good, only question is re special handling of the bool fields.

jharveyb commented:
Added the info command with a few flags to keep the output human-readable in the case of many configs present.

Tested manually on mainnet to display the default config correctly, successfully change settings individually, and reject universe sync for a specific asset.

Fixed an issue with how default configs are applied when we have some custom configs.
